modules/javafx.graphics/src/main/java/com/sun/javafx/application/preferences/PlatformPreferences.java

```diff
@@ -0,0 +1,237 @@
+    @Override
+    public void removeListener(MapChangeListener<? super String, ? super Object> listener) {
+        mapChangeListeners.remove(listener);
+    }
+
+    @Override
+    public <T> Optional<T> getValue(String key, Class<T> type) {
+        Objects.requireNonNull(key, "key cannot be null");
+        Objects.requireNonNull(type, "type cannot be null");
+        Object value = effectivePreferences.get(key);
+
+        if (value == null) {
+            return Optional.empty();
+        }
+
+        if (type.isInstance(value)) {
+            @SuppressWarnings("unchecked")
+            T v = (T)value;
+            return Optional.of(v);
+        }
+
+        throw new IllegalArgumentException(
+            "Incompatible types: requested = " + type.getName() +
+            ", actual = " + value.getClass().getName());
+    }
+
+    @Override
+    public Optional<Integer> getInteger(String key) {
+        return getValue(key, Integer.class);
+    }
+
+    @Override
+    public Optional<Double> getDouble(String key) {
+        return getValue(key, Double.class);
+    }
+
+    @Override
+    public Optional<Boolean> getBoolean(String key) {
+        return getValue(key, Boolean.class);
+    }
+
+    @Override
+    public Optional<String> getString(String key) {
+        return getValue(key, String.class);
+    }
+
+    @Override
+    public Optional<Color> getColor(String key) {
+        return getValue(key, Color.class);
+    }
+
+    @Override
+    public Optional<Paint> getPaint(String key) {
+        return getValue(key, Paint.class);
+    }
+
+    @Override
+    public ReadOnlyObjectProperty<Appearance> appearanceProperty() {
+        return properties.appearanceProperty();
+    }
+
+    @Override
+    public Appearance getAppearance() {
+        return properties.getAppearance();
+    }
+
+    @Override
+    public ReadOnlyObjectProperty<Color> backgroundColorProperty() {
+        return properties.backgroundColorProperty();
+    }
+
+    @Override
+    public Color getBackgroundColor() {
+        return properties.getBackgroundColor();
+    }
+
+    @Override
+    public ReadOnlyObjectProperty<Color> foregroundColorProperty() {
+        return properties.foregroundColorProperty();
+    }
+
+    @Override
+    public Color getForegroundColor() {
+        return properties.getForegroundColor();
+    }
+
+    @Override
+    public ReadOnlyObjectProperty<Color> accentColorProperty() {
+        return properties.accentColorProperty();
+    }
+
+    @Override
+    public Color getAccentColor() {
+        return properties.getAccentColor();
+    }
+
+    /**
+     * Updates this map of preferences with a new set of platform preferences.
+     * The specified preferences may include all available preferences, or only the changed preferences.
+     *
+     * @param preferences the new preference mappings
+     * @throws NullPointerException if {@code preferences} is {@code null}
+     */
+    public void update(Map<String, Object> preferences) {
```

Comment:
I want to make sure I understand the exact details of the update rules. Is this correct?

For each key in `preferences`:
* if it exists in `effectivePreferences`:
    * if the values are the same, there is no change (`preferences` contains all mapping and not just the changes)
    * if the value in `preferences` is `null`, then this is a removal operation (because the value of `effectivePreferences` can't be `null`)
    * if the value in `preferences` is not `null` and not the same, then this is an update operation
* if it doesn't exist in `effectivePreferences`:
    * if the value in `preferences` is not `null`, then this is an add operation
    * can the value in `preferences` be `null`?

Reply:
That's correct.

> can the value in preferences be `null`?

It could be `null` (signalling a removal), but then that's a no-op if the value wasn't present in `effectivePreferences`.

Reply:
So now I'm wondering why the need for the new type `ChangedValue` when you're already accumulating changes with `SimpleChange`.

The update method can be something like

```java
    public void update(Map<String, Object> preferences) {
        var changes = new ArrayList<SimpleChange<String, Object>>();
        preferences.forEach((key, newValue) -> {
            var oldValue = effectivePreferences.get(key);
            // oldValue can't be null, so no NPE
            if (Objects.deepEquals(oldValue, newValue)) {
                return;
            }

            var change = new SimpleChange<>(this);
            if (oldValue != null && newValue == null) {
                change.setRemoved(key, oldValue);
                effectivePreferences.remove(key);
            } else if (oldValue != null && newValue != null) {
                change.setPut(key, oldValue, newValue);
                effectivePreferences.put(key, newValue);
            } else {
                change.setAdded(key, newValue);
                effectivePreferences.put(key, newValue);
            }
            changes.add(change);
        });

        if (!changes.isEmpty()) {
            properties.update(changes, platformKeyMappings);
            fireValueChangedEvent(changes);
        }
    }
```

with `fireValueChangedEvent` being
```java
    private void fireValueChangedEvent(List<SimpleChange<String, Object>> changes) {
        invalidationListeners.forEach(listener -> listener.invalidated(this));
        changes.forEach(change -> mapChangeListeners.forEach(listener -> listener.onChanged(change)));
    }
```
Saves the multiple iterations and the need for a new type.

Then `PreferenceProperties::update` can be adjusted:
```java
    public void update(List<SimpleChange<String, Object>> changes, Map<String, String> platformKeyMappings) {
        for (SimpleChange<String, Object> change : changes) {
        ...
```

I didn't test this, it's just an insight.

Reply:
The main reason for the new record `ChangedValue` is to modularize the code for easier reuse. At the moment, only `PlatformPreferences` uses it, but a follow-up enhancement will introduce a second class `ApplicationPreferences`, which will also use it to compute the effective changes. At that point, we'd either have to duplicate some code, or factor it out anyway.